rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -1979,6 +1979,10 @@ static rmw_ret_t rmw_take_int(
         memcpy(
           message_info->publisher_gid.data, &info.publication_handle,
           sizeof(info.publication_handle));
+        message_info->source_timestamp = info.source_timestamp;
+        if(rcutils_system_time_now(&(message_info->received_timestamp)) != RCUTILS_RET_OK) {
```

Comment:
We shouldn't just silently ignore a failure to take the current time... This should fail the function, in my opinion.

Reply:
I'm now returning an error in this case.

However, this leads to the somewhat counter-intuitive situation that the message has been taken from DDS (and this is marked properly), but the return code is an error. AFAICT, the user needs to free the message in this case. We should check the places where this is used to ensure that actually happens.

Reply:
The user always needs to destruct (destructor in C++ and fini in C) and free (opposite of malloc or delete, opposite of new) the message, even if take fails...

So I don't think that's an issue.

Taking the data but not giving it to the user is is unfortunate, but we can have a mixed result, like:

> If the error code `RMW_TAKE_SUCCEEDED_BUT_META_DATA_FAILED` is returned, then you should still check `taken` to see if data was taken, but the data in the message info may be incomplete.

Then the user can decide what to do in that case.

I think it is unlikely that this function fails, but in that event, I think failing with a return code (even if the post condition w.r.t. the message being taken is inconvenient) is better than silently ignoring it and not filling the meta data.

Reply:
To be clear, I think returning `RMW_RET_ERROR` for now is fine. I was just suggesting how we could improve the error reporting/documentation if we felt we needed to do so. I don't feel strongly that is the case, however, since I think this function failing is pretty unlikely in practice.

Reply:
I agree on the error case, but I'm thinking we should always return 0 here, to signal that the received timestamp is not really supported.

Regarding the error, I think you're right. Aleast on POSIX, CLOCK_REALTIME is guaranteed to always be present, so in practice this can't fail (at least not with any of its documented error codes). Moreover, the C++ version of this, std::chrono::system_clock::now, is even declared noexcept.

However, obtaining the received_timestamp at the point of taking the message is not representative of when the message is really received. We really need the DDS implementation to support an internal received timestamp (like Fast-RTPS 1.10 does). As long as we cannot rely on all of the implementations to support this, we cannot really use the received timestamp. The only way to distinguish that would be that those implementations that do not support it return 0 to signal this.

Unless you object to this reasoning, I would replace this with 0.